libraries/apollo-ast/src/commonMain/kotlin/com/apollographql/apollo3/ast/introspection/meta_introspection_reader.kt

```
@@ -0,0 +1,156 @@
+  val introspection = this.toMIntrospection()
+  return buildSet {
+    val types = introspection.data.__schema.types
+    types.firstOrNull { it.name == "__Schema" }?.let { schema ->
+      if (schema.fields.orEmpty().any { it.name == "description" }) {
+        add(SchemaDescription)
+      }
+    }
+    types.firstOrNull { it.name == "__Type" }?.let { type ->
+      if (type.fields.orEmpty().any { it.name == "specifiedByURL" }) {
+        add(TypeSpecifiedByURL)
+      }
+      if (type.fields.orEmpty().any { it.name == "isOneOf" }) {
+        add(TypeIsOneOf)
+      }
+      type.fields.orEmpty().firstOrNull{ it.name == "inputFields" }?.let { inputFields ->
+        if (inputFields.args.any { it.name == "includeDeprecated" }) {
+          add(TypeInputFieldsIncludeDeprecated)
+        }
+      }
+    }
+    types.firstOrNull { it.name == "__Directive" }?.let { directive ->
+      if (directive.fields.orEmpty().any { it.name == "isRepeatable" }) {
+        add(DirectiveIsRepeatable)
+      }
+      directive.fields.orEmpty().firstOrNull{ it.name == "args" }?.let { args ->
+        if (args.args.any { it.name == "includeDeprecated" }) {
+          add(DirectiveArgsIncludeDeprecated)
+        }
+      }
+    }
+    types.firstOrNull { it.name == "__Field" }?.let { field ->
+      field.fields.orEmpty().firstOrNull{ it.name == "args" }?.let { args ->
+        if (args.args.any { it.name == "includeDeprecated" }) {
+          add(FieldArgsIncludeDeprecated)
+        }
+      }
+    }
+    types.firstOrNull { it.name == "__InputValue" }?.let { inputValue ->
+      if (inputValue.fields.orEmpty().any { it.name == "isDeprecated" }) {
+        add(InputValueIsDeprecated)
+      }
+      if (inputValue.fields.orEmpty().any { it.name == "deprecationReason" }) {
+        add(InputValueDeprecatedReason)
+      }
+    }
+  }
+}
+
+private val json = Json {
+  // be robust to errors: [] keys
+  ignoreUnknownKeys = true
```

What happens if a field errors (very unlikely to happen for an introspection query but 🤷 )?  Should we output an error instead?

---

If you get an error here, something is very wrong, so it would make sense to fail early I think.